File: core/UnivaluedBinaryTree/UnivaluedBinaryTree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isUnivalTree(root: TreeNode) -> bool:
    stack, cur, remain = [], root, -1
    while stack or cur:
        while cur:
            stack.append(cur)
            cur = cur.left
        cur = stack.pop()
        if remain != -1 and remain != cur.val:
            return False
        remain = cur.val
        cur = cur.right
    return True

File: core/UnivaluedBinaryTree/UnivaluedBinaryTree.java
package com.dbc;

import java.util.Stack;

public class UnivaluedBinaryTree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public boolean isUnivalTree(TreeNode root) {
        Stack<TreeNode> stack = new Stack<>();
        TreeNode curNode = root;
        int remain = -1;
        while (!stack.isEmpty() || curNode != null) {
            while (curNode != null) {
                stack.add(curNode);
                curNode = curNode.left;
            }
            curNode = stack.pop();
            if (remain != -1 && remain != curNode.val) return false;
            remain = curNode.val;
            curNode = curNode.right;
        }
        return true;
    }
}
